File: core/jvm/src/main/java/zio/internal/FiberSet.java
/*
 * Copyright 2022-2024 John A. De Goes and the ZIO Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package zio.internal;

import java.lang.ref.ReferenceQueue;
import java.lang.ref.WeakReference;
import java.util.AbstractSet;
import java.util.Iterator;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * A high-performance, Loom-friendly concurrent weak set specialized for Fiber.Runtime.
 * This implementation avoids global synchronization and carrier thread pinning
 * by using ConcurrentHashMap with specialized identity-based WeakKey wrappers.
 */
public final class FiberSet<A> extends AbstractSet<A> {
    private final ConcurrentHashMap<WeakKey<A>, Boolean> map = new ConcurrentHashMap<>();
    private final ReferenceQueue<A> queue = new ReferenceQueue<>();
    private final AtomicInteger size = new AtomicInteger();

    private static final class WeakKey<A> extends WeakReference<A> {
        private final int hashCode;

        WeakKey(A referent, ReferenceQueue<A> q) {
            super(referent, q);
            this.hashCode = System.identityHashCode(referent);
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) return true;
            if (obj instanceof WeakKey) {
                Object r1 = get();
                Object r2 = ((WeakKey<?>) obj).get();
                return r1 != null && r1 == r2;
            }
            return false;
        }

        @Override
        public int hashCode() {
            return hashCode;
        }
    }

    private void expungeStaleEntries() {
        WeakKey<A> key;
        while ((key = (WeakKey<A>) queue.poll()) != null) {
            if (map.remove(key) != null) {
                size.decrementAndGet();
            }
        }
    }

    @Override
    public boolean add(A a) {
        if (a == null) throw new NullPointerException();
        expungeStaleEntries();
        if (map.putIfAbsent(new WeakKey<>(a, queue), Boolean.TRUE) == null) {
            size.incrementAndGet();
            return true;
        }
        return false;
    }

    @SuppressWarnings("unchecked")
    @Override
    public boolean remove(Object o) {
        if (o == null) return false;
        expungeStaleEntries();
        if (map.remove(new WeakKey<>((A) o, null)) != null) {
            size.decrementAndGet();
            return true;
        }
        return false;
    }

    @Override
    public int size() {
        expungeStaleEntries();
        return size.get();
    }

    @SuppressWarnings("unchecked")
    @Override
    public boolean contains(Object o) {
        if (o == null) return false;
        return map.containsKey(new WeakKey<>((A) o, null));
    }

    @Override
    public Iterator<A> iterator() {
        expungeStaleEntries();
        final Iterator<WeakKey<A>> it = map.keySet().iterator();
        return new Iterator<A>() {
            private A next;

            @Override
            public boolean hasNext() {
                while (it.hasNext()) {
                    next = it.next().get();
                    if (next != null) return true;
                }
                return false;
            }

            @Override
            public A next() {
                if (next == null && !hasNext()) {
                    throw new java.util.NoSuchElementException();
                }
                A val = next;
                next = null;
                return val;
            }

            @Override
            public void remove() {
                it.remove();
                size.decrementAndGet();
            }
        };
    }

    @Override
    public void clear() {
        map.clear();
        size.set(0);
        while (queue.poll() != null);
    }
}

File: core/jvm/src/main/scala/zio/internal/FiberSet.scala
/*
 * Copyright 2022-2024 John A. De Goes and the ZIO Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package zio.internal

import zio.stacktracer.TracingImplicits.disableAutoTrace

import java.lang.ref.{ReferenceQueue, WeakReference}
import java.util.concurrent.ConcurrentHashMap
import java.util.{AbstractSet, Iterator => JIterator}

/**
 * A Loom-friendly, concurrent, weak-reference set for tracking [[FiberRuntime]]
 * children.
 *
 * Replaces the previous
 * `Collections.synchronizedSet(Collections.newSetFromMap(new WeakHashMap))` on
 * the JVM with a [[ConcurrentHashMap]]-backed implementation that avoids the
 * global ''synchronized'' monitor. On Project Loom, a ''synchronized'' block
 * pins the carrier thread and prevents virtual-thread scheduling; this
 * implementation uses only the internal striped locks of [[ConcurrentHashMap]],
 * which are Loom-compatible.
 *
 * ==Design==
 *
 * Elements are stored as [[IdentityWeakKey]] instances:
 * {{{
 *   ConcurrentHashMap[ IdentityWeakKey[A], java.lang.Boolean ]
 * }}}
 * Each key holds a [[WeakReference]] to the element and records the element's
 * [[System.identityHashCode]] at insertion time. When the GC collects the
 * element the key is enqueued in a [[ReferenceQueue]]; the next call to
 * [[add]], [[size]], [[isEmpty]], or [[iterator]] drains that queue and removes
 * the corresponding dead keys from the map ([[expunge]]).
 *
 * ==Thread safety==
 *
 * All mutations delegate to [[ConcurrentHashMap]]; individual operations are
 * linearisable. [[expunge]] is not synchronized so, under contention, two
 * threads may both try to remove the same dead key — this is harmless because
 * [[ConcurrentHashMap.remove]] is idempotent.
 *
 * ==Eventual consistency of [[isEmpty]]==
 *
 * Like [[FiberMailbox.isEmpty]], this method is ''eventually consistent'':
 * a concurrent [[add]] that has not yet completed may cause a transient false
 * positive. Callers (i.e. [[FiberRuntime]]) must tolerate false negatives.
 */
private[zio] final class FiberSet[A <: AnyRef] extends AbstractSet[A] {

  private val backing  = new ConcurrentHashMap[IdentityWeakKey[A], java.lang.Boolean]()
  private val refQueue = new ReferenceQueue[A]()

  // -------------------------------------------------------------------------
  // AbstractSet contract
  // -------------------------------------------------------------------------

  override def add(a: A): Boolean = {
    if (a eq null) throw new NullPointerException("FiberSet does not accept null elements")
    expunge()
    backing.put(new IdentityWeakKey[A](a, refQueue), java.lang.Boolean.TRUE) eq null
  }

  override def size(): Int = {
    expunge()
    backing.size()
  }

  override def isEmpty: Boolean = {
    expunge()
    backing.isEmpty
  }

  override def iterator(): JIterator[A] = {
    expunge()
    new JIterator[A] {
      private val inner  = backing.keySet().iterator()
      private var _next  = fetchNext()

      private def fetchNext(): A = {
        while (inner.hasNext) {
          val v = inner.next().get()
          if (v ne null) return v
          else inner.remove()
        }
        null.asInstanceOf[A]
      }

      override def hasNext: Boolean = _next ne null
      override def next(): A = {
        if (_next eq null) throw new java.util.NoSuchElementException
        val result = _next
        _next = fetchNext()
        result
      }
    }
  }

  // -------------------------------------------------------------------------
  // GC cleanup
  // -------------------------------------------------------------------------

  /**
   * Drains GC'd entries from the [[ReferenceQueue]] and removes them from the
   * backing map. Called automatically before any structural operation.
   */
  private def expunge(): Unit = {
    var ref = refQueue.poll()
    while (ref ne null) {
      backing.remove(ref)
      ref = refQueue.poll()
    }
  }
}

/**
 * An identity-based [[WeakReference]] key for use inside [[ConcurrentHashMap]].
 *
 * Two [[IdentityWeakKey]] instances are ''equal'' iff their referents are the
 * same object (reference equality), falling back to [[AnyRef]] reference
 * equality when one or both referents have been GC'd (so that dead keys can
 * still be removed via [[ConcurrentHashMap.remove]] from [[FiberSet.expunge]]).
 *
 * The hash code is fixed at construction time to the identity hash of the
 * referent so that the key remains usable after the referent is collected.
 */
private[internal] final class IdentityWeakKey[A <: AnyRef](ref: A, queue: ReferenceQueue[A])
    extends WeakReference[A](ref, queue) {

  private val _hash: Int = System.identityHashCode(ref)

  override def hashCode(): Int = _hash

  override def equals(obj: Any): Boolean = obj match {
    case other: IdentityWeakKey[_] =>
      val a = get()
      val b = other.get()
      if ((a ne null) && (b ne null)) a eq b  // both alive: compare by identity
      else this eq other                        // at least one dead: fall back to ref equality
    case _ => false
  }
}
